l5kit/l5kit/simulation/dataset.py

```diff
@@ -46,19 +46,34 @@ def __init__(self, scene_dataset_batch: Dict[int, EgoDataset], sim_cfg: Simulati
         if not len(scene_dataset_batch):
             raise ValueError("can't build a simulation dataset with an empty batch")
         self.scene_dataset_batch: Dict[int, EgoDataset] = scene_dataset_batch
+        self.sim_cfg = sim_cfg
 
-        self._len = min([len(scene_dt.dataset.frames) for scene_dt in self.scene_dataset_batch.values()])
+        # we must limit the scenes to the part which will be simulated
+        # we cut each scene so that it starts from there and ends after `num_simulation_steps`
+        start_frame_idx = self.sim_cfg.start_frame_index
+        if self.sim_cfg.num_simulation_steps is None:
+            end_frame_idx = min([len(scene_dt.dataset.frames) for scene_dt in self.scene_dataset_batch.values()])
+        else:
+            end_frame_idx = start_frame_idx + self.sim_cfg.num_simulation_steps
 
-        self.sim_cfg = sim_cfg
+        for scene_idx in scene_dataset_batch:
+            zarr_dt = self.scene_dataset_batch[scene_idx].dataset
+            self.scene_dataset_batch[scene_idx].dataset = get_frames_subset(zarr_dt, start_frame_idx, end_frame_idx)
+
+            # this is the only stateful field we need to change for EgoDataset, it's used in bisect
+            frame_index_ends = self.scene_dataset_batch[scene_idx].dataset.scenes["frame_index_interval"][:, 1]
+            self.scene_dataset_batch[scene_idx].cumulative_sizes = frame_index_ends
+
+        self._len = min([len(scene_dt.dataset.frames) for scene_dt in self.scene_dataset_batch.values()])
```

Comment:
This code is repeating the same as line 55 above. You can create a method for it.

Reply:
sure